python/mage/tgn/definitions/tgn.py

```diff
@@ -0,0 +1,1213 @@
+
+            curr_mapped_nodes = np.array([mapping[(v, t)] for (v, t) in nodes])
+
+            keys = aggregate_unsqueeze
+            values = aggregate_unsqueeze
+            query_concat = torch.concat(
+                (
+                    out[curr_mapped_nodes],
+                    time_encoder_zeros.repeat(len(curr_mapped_nodes), 1),
+                ),
+                dim=1,
+            )
+            query = torch.unsqueeze(query_concat, dim=0)
+
+            attn_out, _ = self.multi_head_attentions[k](
+                query=query, key=keys, value=values
+            )
+
+            attn_out = torch.squeeze(attn_out)
+
+            concat_neigh_out = torch.cat((out[curr_mapped_nodes], attn_out), dim=1)
+            out = self.mlps[k](concat_neigh_out)
+        return out
+
+    def _aggregate(self, features, rows, nodes, mapping, edge_features, time_features):
+        assert len(nodes) == len(rows)
+        mapped_rows = [
+            np.array([mapping[(vi, ti)] for (vi, ti) in row]) for row in rows
+        ]
+
+        out = torch.rand(len(nodes), self.num_neighbors * self.key_dim)
+
+        # row represents list of indexes of "current neighbors" of edge_features on i-th index
+        for i, row in enumerate(mapped_rows):
+            features_curr = features[row][:]
+            edge_feature_curr = edge_features[i][:]
+            time_feature_curr = time_features[i][:]
+
+            # shape(1, num_neighbors * (embedding_dim + edge_features_dim + time_encoding_dim)
+            # after doing concatenation on columns side, reshape to have 1 row
+            aggregate = torch.concat(
+                (features_curr, edge_feature_curr, time_feature_curr), dim=1
+            ).reshape((1, -1))
+
+            out[i, :] = aggregate
+
+        return out
+
+
+def get_message_function_type(message_function_type: MessageFunctionType):
+    if message_function_type == MessageFunctionType.MLP:
+        return MessageFunctionMLP
+    elif message_function_type == MessageFunctionType.Identity:
+        return MessageFunctionIdentity
+    else:
+        raise Exception(
+            f"Message function type {message_function_type} not yet supported."
+        )
+
+
+def get_memory_updater_type(memory_updater_type: MemoryUpdaterType):
+    if memory_updater_type == MemoryUpdaterType.GRU:
+        return MemoryUpdaterGRU
+
+    elif memory_updater_type == MemoryUpdaterType.RNN:
+        return MemoryUpdaterRNN
+    else:
+        raise Exception(f"Memory updater type {memory_updater_type} not yet supported.")
+
+
+def get_message_aggregator_type(message_aggregator_type: MessageAggregatorType):
+    if message_aggregator_type == MessageAggregatorType.Mean:
+        return MeanMessageAggregator
+
+    elif message_aggregator_type == MessageAggregatorType.Last:
+        return LastMessageAggregator
+
+    else:
+        raise Exception(
+            f"Message aggregator type {message_aggregator_type} not yet supported."
+        )
```

Comment:
Instances of TemporalNetworks might be in different file :D

Reply:
Yes, it makes sense 😄